RavenDB-19421 Will send conflict-resolved document to the other side …

ayende opened the pull request:
…to ensure that all nodes in the cluster will have the same change vector for the documents

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19421 

### Additional description

When we resolve a conflict using "it's the same thing" model, we need to send it to the other side anyway, to ensure that the change vectors would be the same afterward.


### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

ayende committed:
RavenDB-19421 Will send conflict-resolved document to the other side to ensure that all nodes in the cluster will have the same change vector for the documents

ayende committed:
RavenDB-18647 Adding test for conflict on attachment details that is properly resolved

ayende commented:
@shiranshalom Please look especially at the test in `ConflictOfAttachmentAndDocument3StoresDifferentLastModifiedOrder`.
It is passing for me after the change, but I want to verify that this is testing the same thing on your end.

Another thing to note is that a lot of your changes here aren't there:
https://github.com/ayende/ravendb/commit/4e7f13cfe580bf209388c452acd5f74c85d67757

karmeli87 reviewed (approved):
LGTM

@ayende I think some of the tests were failing when batch size was set to 1.

ppekrol commented on src/Raven.Server/Documents/Handlers/ReplicationHandler.cs:
```diff
@@ -41,6 +43,60 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
             }
         }
+        
+        [RavenAction("/databases/*/replication/all-items", "GET", AuthorizationStatus.ValidUser, EndpointType.Read)]
```
Looks like a debug endpoint?
Maybe `/databases/*/debug/replication/all-items` ?

ayende replied:
Done, explicitly not setting `IsDebugInformationEndpoint` to avoid adding this to the debug package

ayende committed:
RavenDB-18647 Fixing tests to add the Resolved flag. Making sure that the debug endpoint is in the right /debug/ path

shiranshalom committed:
RavenDB-19421 - add tests

ayende committed:
RavenDB-19421
- When resolving a conflict using resolve to latest, will use the latest version as the modified date for the resolved document
- When resolving a conflict on documents attachment, will merge the change vectors of the attachments to ensure the same change vector is throughout the cluster
- Properly handle conflict on the same attachment name between nodes and update the change vector

ppekrol commented on src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:
```diff
@@ -477,29 +481,44 @@ public DocumentConflict ResolveToLatest(List<DocumentConflict> conflicts)
             using var scope = Slice.External(context.Allocator, resolved.LowerId, out var lowerId);
-            var storageAttachmentsDetails = _database.DocumentsStorage.AttachmentsStorage.GetAttachmentDetailsForDocument(context, lowerId);
-            List<AttachmentName> resolvedAttachmentsMetadata = null;
+            var resolvedAttachmentsMetadata = AttachmentsStorage.GetAttachmentsFromDocumentMetadata(resolved.Doc)
```
this can be moved down? 

since the purpose of that is to seek, then maybe it would be better to build a dict here

```
struct AttachmentMetadataKey
{
    string Name;
    string Hash;
    string ContentType;
}

Dictionary<AttachmentMetadataKey, AttachmentDetails>
```

instead of using `.Any` each time?

ayende replied:
The likely scenario is that you have, at max, a few dozens of attachments.
And conflicts are rare, I don't think it would be worth it

arekpalinski commented:
Note that https://jenkins.hibernatingrhinos.com/job/PR_Tests/5227/testReport/junit/SlowTests.Client.Attachments/AttachmentsReplication/ConflictOfAttachmentAndDocument3Stores/ is still failing

ayende commented:
I wonder if this is a timeout issue, basically. Because _that_ was never a failure mode for the test in question.
The test would fail because of bad data due to the setup, not fail to replicate entirely.

ayende commented:
At any rate, I'll try running this locally in a loop and see where it leads

ayende committed:
RavenDB-19421
- Defer the deletion of attachment streams to the end of the transaction, to allow for a single tx that removes then add the stream
- Adding a test that shows failure with single step transactions and attachment removal

ayende committed:
RavenDB-19421
- If we get an attachment tombstone in a replication batch, without the associated document, we'll update the document to reflect the removed attachment immediately, instead of waiting for the document change
- Sending last modified ticks for attachment tombstones as well during replication

ayende commented:
@shiranshalom @karmeli87 please take a look

shiranshalom reviewed (approved):
LGTM!

karmeli87 commented on src/Raven.Server/Documents/Handlers/ReplicationHandler.cs:
```diff
@@ -41,6 +43,60 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+
+                foreach (ReplicationBatchItem item in ReplicationDocumentSender.GetReplicationItems(Database, context, etag, stats, false).Take(pageSize))
+                {
+                    var djv = new DynamicJsonValue
+                    {
+                        [nameof(item.ChangeVector)] = item.ChangeVector,
+                        [nameof(item.Etag)] = item.Etag,
+                        [nameof(item.Type)] = item.Type,
+                        [nameof(item.TransactionMarker)] = item.TransactionMarker,
+                    };
+                    switch (item)
+                    {
+                        case DocumentReplicationItem i:
+                            djv[nameof(i.Collection)] = i.Collection;
+                            djv[nameof(i.Id)] = i.Id;
+                            djv[nameof(i.Flags)] = i.Flags;
+                            break;
+                        case AttachmentReplicationItem a:
+                            djv[nameof(a.Base64Hash)] = a.Base64Hash.ToString();
+                            djv[nameof(a.Name)] = a.Name.ToString();
+                            djv[nameof(a.ContentType)] = a.ContentType.ToString();
+                            break;
+                    }
```
this is handy, let's maybe add `ToJson` to the `ReplicationBatchItem`

ayende replied:
That is a bigger change that we have to do right now. 
And a virtual method call is likely slower than this FWIW

ayende replied:
Done

karmeli87 commented on src/Raven.Server/Documents/Handlers/ReplicationHandler.cs:
```diff
@@ -41,6 +43,60 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+        public async Task GetAllItems()
+        {
+            var etag = GetLongQueryString("etag", required: false) ?? 0L;
+            var pageSize = GetPageSize();
```
it is easy not to provide it here, which might be bad during a debug session.

ayende replied:
Why? `&pageSize=5500` as a query string argument

karmeli87 commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1179,6 +1189,14 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
+                                var keySpan = attachmentTombstone.Key.AsReadOnlySpan();
+                                int endOfDocumentId = keySpan.IndexOf(SpecialChars.RecordSeparator);
+                                Debug.Assert(endOfDocumentId != -1);
+                                string documentId = Encoding.UTF8.GetString(keySpan[..endOfDocumentId]);// will throw if no separator found 
```
so this will end up in a replication loop, right? is it a valid scenario?

ayende replied:
It will _not_ be a replication loop. It will send this once, yes.
It's an update (merged with the old change vector), it will be sent to the other side, where it will be _older_ than the previous version, and then discarded.

karmeli87 commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1422,6 +1440,30 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
+                            {   
+                                // have to load the full document
+                                doc = context.DocumentDatabase.DocumentsStorage.Get(context, docId, fields: DocumentFields.All, throwOnConflict: false);
+                                long lastModifiedTicks = Math.Max(modifiedTicks, doc.LastModified.Ticks); // old versions may send with 0 in the tombstone ticks
```
the `LastModified` is a new addition to the schema?

karmeli87 replied:
if we have a conflict should we generate a new `ticks` ?

ayende replied:
No, that is something we always had, we just never sent it over the wire.
The `Math.Max` there is to handle the older versions.

This isn't working with conflicts, since then `doc` would be null, and the operation would fix itself on conflict resolve

ppekrol replied:
What will happen if one of the nodes will not send it? asking about Interversion compatibility

ayende replied:
Then we will default to the last modified on the doc, which is still okay (will resolve the conflict to the other version when it arrives)

We are already sending `0` over the wire, so this is fine

ayende replied:
They are already sending 0

karmeli87 reviewed (approved):
Mostly cosmetics, can you think of any implication on backward compatibility that we should care?

ppekrol commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1179,6 +1189,14 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
+                                var keySpan = attachmentTombstone.Key.AsReadOnlySpan();
+                                int endOfDocumentId = keySpan.IndexOf(SpecialChars.RecordSeparator);
+                                Debug.Assert(endOfDocumentId != -1);
+                                string documentId = Encoding.UTF8.GetString(keySpan[..endOfDocumentId]);// will throw if no separator found 
```
sth that worries me is that we have a piece of code like that here, don't we have some sort of a helper class that centralizes this? Can't we have some AttachmentKeyHelper or sth like that?

ayende replied:
Done

ayende committed:
RavenDB-19421 Refactor debug information generation for replication items
